Remove helper method frame from Profiler hooks

davidwrighton opened the pull request:
Remove helper method frame and update the CONTRACTs to suit the new behavior. This WILL change the profiler behavior in that a set of apis in the Profiler api surface will not fail immediately on any attempt to use them. However, the set of apis which change are I believe not safe to call within a ProfilerEnter/Leave hook as an actual triggered GC will cause potential memory corruption if the arguments/return value of the function have any GC pointers.

This be done without the tweaks to remove the CodeVersionManager lock, but in general, I'd like to get rid of that lock from most of the entire runtime. I welcome opinions as to whether I should try to hold the CodeVersionManager lock into another change focussed on that.

davidwrighton committed:
Remove helper method frame from Profiler hooks

davidwrighton committed:
Adjust lock checking around IL and Native Code versioning
- Now, you can iterate the lists, and read the various rejit id's safely

This enables a few of the apis on the profiler to be safe to call while in a funciton enter/leave hook. From my analysis these functions were previously safe to call and did not assert, but only because of the presence of a HELPER_METHOD_FRAME. Now, they should simply be safe to call and the erroneous failure to assert has been addressed

davidwrighton committed:
Merge branch 'main' of github.com:dotnet/runtime into RemoveHelperMethodFrameFromProfilerHooks

davidwrighton committed:
Fix compile issues on Unix platforms

davidwrighton committed:
Make ProfToEEInterfaceImpl::GetAppDomainInfo safe to call without a helper method frame
- Rework AppDomain::GetFriendlyName logic so that it is always safe for use here.

davidwrighton committed:
Remove another place where we have problematic contracts

davidwrighton commented:
This be done without the tweaks to remove the CodeVersionManager lock, but in general, I'd like to get rid of that lock from most of the entire runtime. I welcome opinions as to whether I should try to hold the CodeVersionManager lock into another change focussed on that.

AaronRobinsonMSFT commented on src/coreclr/vm/appdomain.cpp:
```diff
@@ -2876,30 +2860,37 @@ void AppDomain::SetFriendlyName(LPCWSTR pwzFriendlyName, BOOL fDebuggerCares/*=T
     {
-        THROWS;
-        if (GetThreadNULLOk()) {GC_TRIGGERS;} else {DISABLED(GC_NOTRIGGER);}
+        NOTHROW;
```
This isn't correct. If `m_friendlyName` isn't in UTF16 format, then `GetUnicode()` will throw.

davidwrighton replied:
Done.

davidwrighton committed:
Get rid of SString as a field on the AppDomain type for friendly name
- Get rid of GetFriendlyNameForLogging, as its difference from GetFriendlyName is immaterial at the only time its ever called

davmason commented:
> This WILL change the profiler behavior in that a set of apis in the Profiler api surface will not fail immediately on any attempt to use them. However, the set of apis which change are I believe not safe to call within a ProfilerEnter/Leave hook as an actual triggered GC will cause potential memory corruption if the arguments/return value of the function have any GC pointers.

Is "not fail immediately" a typo for "now fail immediately"? As written this worries me, the whole profiler ELT hook concept is not well understood by the community at large, and very poorly documented by us. We shouldn't be introducing new ways for profilers to silenty fail in weird ways.

davidwrighton committed:
Merge branch 'main' of github.com:dotnet/runtime into RemoveHelperMethodFrameFromProfilerHooks

noahfalk commented on src/coreclr/vm/codeversion.h:
```diff
@@ -287,28 +287,28 @@ class NativeCodeVersionNode
 
-    ReJITID m_parentId;
+    DAC_IGNORE(const) ReJITID m_parentId;
     PTR_NativeCodeVersionNode m_pNextMethodDescSibling;
```
There is a comment below saying m_pNextILVersionNode is never modified after being added to the list. I'm guessing the same thing holds here and we could add a similar comment?

noahfalk reviewed (approved):
A few small suggestions inline, looks like a nice improvement overall. I'm happy to see that lock diminished 👍

noahfalk commented:
> I welcome opinions as to whether I should try to hold the CodeVersionManager lock into another change focussed on that.

This change seemed like a good scope as-is. I wouldn't add more lock refactoring to this PR if that is what you were asking about.

davidwrighton commented:
@davmason, we have logic which attempts to replicate the behavior of our contracts system, but with runtime failures. This is the logic which calls `PROFILER_TO_CLR_ENTRYPOINT_SYNC_EX` with `kP2EETriggers` as one of the flags. The behavior will be to return `CORPROF_E_UNSUPPORTED_CALL_SEQUENCE`. This change makes calls to the following functions throw if used from an profile enter/leave event.

`GetAppDomainsContainingModule`,
`GetILFunctionBodyAllocator`,
`SetILFunctionBody`,
`SetILInstrumentedCodeMap`,
`ForceGC`
`GetClassFromToken`,
`GetClassFromTokenAndTypeArgs`,
`GetFunctionFromTokenAndTypeArgs`,
`GetAppDomainInfo`
`RequestReJITWithInliners`,
`GetReJITIDs`,
`RequestReJIT`,
`RequestRevert`,
`EnumJITedFunctions2`,
`EnumModules`,
`EnumNgenModuleMethodsInliningThisMethod`,
`ApplyMetaData`,

This is unfortunate, so I went through the work to make some of the more likely used apis reliably succeed.

However, I can't fix all of them (although I could fix *some* of them). We also have the problem that we already had an unpredictable failure case, where if we actually hit a GC suspend point during an Enter/Leave hook, we will corrupt memory. This change makes more apis fail, but the ones that work should reliably work unlike the previous model where some stuff worked but only if the customer didn't get happen to get unlucky.

tommcdon reviewed (approved):
Thank you!

davmason reviewed (approved):
LGTM, I am a little worried about the implications for the profiler but doing this at the beginning of 10 should help with that

davidwrighton commented on src/coreclr/debug/ee/functioninfo.cpp:
```diff
@@ -1053,15 +1051,22 @@ void DebuggerJitInfo::SetBoundaries(ULONG32 cMap, ICorDebugInfo::OffsetMapping *
+        }
+        else
+        {
+            _ASSERTE(!"Unexpected rejit state, should be active as there exists a native code version for this IL code version");
```
`ILCodeVersion` goes from state `kStateRequested` to `kStateGettingReJITParameters` to `kStateActive`. We cannot actually run the JIT without the function being active, as we don't actually have any IL. There is no pathway for an `ILCodeVersion` to become inactive. 

It however can be not THE active IL Code version, which is a different concept (which confusingly doesn't require the `ILCodeVersion` to have reached `kStateActive`. Honestly, there are 4 concepts of "active" in the code versioning codebase, all of which mean fairly different things, which makes everything here really confusing to read.

The `DebuggerJitInfo` can only be created after jitting completes, as the constructor for a `DebuggerJitInfo` requires a pointer to the start of the jitted code.

noahfalk replied:
Ah yes, I forgot that RejitState() == kStateActive was distinct from CodeVersionManager::GetActiveILCodeVersion(). And agreed, the naming collision doesn't make that convenient to discuss. Probably not in this PR, but what you do think about renaming kStateActive to something like kILInitialized or kReJITParametersInitialized? I think that would better capture its meaning and reduce overload on the 'active' term.

davidwrighton committed:
Code review feedback

davidwrighton committed:
Merge branch 'main' of github.com:dotnet/runtime into RemoveHelperMethodFrameFromProfilerHooks
